File: tests/test_cli.py
"""CLI smoke tests."""

from __future__ import annotations

from typer.testing import CliRunner

from sbsave.cli import app
from sbsave.savegame import discover_saves

runner = CliRunner()


def test_help():
    result = runner.invoke(app, ["--help"])
    assert result.exit_code == 0
    assert "report" in result.output


def test_version():
    result = runner.invoke(app, ["--version"])
    assert result.exit_code == 0
    assert "sbsave" in result.output


def test_catalog_list():
    result = runner.invoke(app, ["catalog", "list"])
    assert result.exit_code == 0
    assert "nano_suits" in result.output


def test_catalog_check():
    result = runner.invoke(app, ["catalog", "check"])
    assert result.exit_code == 0


def test_report_with_real_save():
    slots = discover_saves()
    if not slots:
        return
    result = runner.invoke(app, ["report", "--category", "cans"])
    assert result.exit_code == 0
    assert "罐子" in result.output


def test_dump_with_real_save():
    slots = discover_saves()
    if not slots:
        return
    result = runner.invoke(app, ["dump"])
    assert result.exit_code == 0
    assert "SBSaveGame" in result.output

File: crates/sbsave-cli/tests/test_cli.rs
use std::process::{Command, Output};

fn sbsave(args: &[&str]) -> Output {
    Command::new(env!("CARGO_BIN_EXE_sbsave"))
        .args(args)
        .output()
        .expect("run sbsave")
}

fn stdout(output: &Output) -> String {
    String::from_utf8_lossy(&output.stdout).into_owned()
}

fn has_local_save() -> bool {
    !sbsave_core::savegame::discover_saves(None).is_empty()
}

#[test]
fn help_lists_report() {
    let output = sbsave(&["--help"]);
    assert!(output.status.success());
    assert!(stdout(&output).contains("report"));
}

#[test]
fn version_prints_name() {
    let output = sbsave(&["--version"]);
    assert!(output.status.success());
    assert!(stdout(&output).contains("sbsave"));
}

#[test]
fn catalog_list_shows_categories() {
    let output = sbsave(&["catalog", "list"]);
    assert!(output.status.success());
    assert!(stdout(&output).contains("nano_suits"));
}

#[test]
fn catalog_check_passes() {
    let output = sbsave(&["catalog", "check"]);
    assert!(output.status.success());
}

#[test]
fn report_with_real_save() {
    if !has_local_save() {
        return;
    }
    let output = sbsave(&["report", "--category", "cans"]);
    assert!(output.status.success());
    assert!(stdout(&output).contains("罐子"));
}

#[test]
fn dump_with_real_save() {
    if !has_local_save() {
        return;
    }
    let output = sbsave(&["dump"]);
    assert!(output.status.success());
    assert!(stdout(&output).contains("SBSaveGame"));
}
